### PR TITLE
Log grads and updates in FSDP

### PR DESCRIPTION
**Pull request by maciek-pioro**
# Description
todo

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary: look at [(scaled_)?(gradient|update)_norm/.*](https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=9a9e4902-9b67-478b-adb4-36df9e86a9c1&detailsTab=metadata&shortId=LLMRANDOM-15007&type=run&path=.)
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/utils/conditional_trainer.py [hunk 1]
```diff
@@ -229,8 +255,58 @@ def calculate_loss_and_gradient(self, processed_batch: LLMBatch):
...
+        if self.will_report_update_norm(step):
+            with FSDP.summon_full_params(self.model, with_grads=False):
+                for name, value in self.model.named_parameters():
+                    eps = 1e-5
+                    update_norm = torch.linalg.norm(
+                        value.detach() - self.model_checkpoint[name]
+                    )
+                    param_norm = torch.linalg.norm(self.model_checkpoint[name])
+                    if self.is_logging_process:
+                        self.logger.report_scalar(
```
Comment (crewtool) [inline research/conditional/utils/conditional_trainer.py:296]: should be part of `will_report_update_norm`

File: research/conditional/utils/conditional_trainer.py [hunk 2]
```diff
@@ -229,8 +255,58 @@ def calculate_loss_and_gradient(self, processed_batch: LLMBatch):
...
-    def _apply_gradient(self):
+    def maybe_report_gradient_norm(self, step: int):
+        if self.logging_interval_heavy > 0 and step % self.logging_interval_heavy == 0:
+            with FSDP.summon_full_params(self.model, with_grads=True):
+                for name, value in self.model.named_parameters():
+                    if value.grad is not None:
+                        eps = 1e-5
+                        grad_norm = torch.linalg.norm(value.grad)
+                        param_norm = torch.linalg.norm(self.model_checkpoint[name])
+                        if self.is_logging_process:
```
Comment (crewtool) [inline research/conditional/utils/conditional_trainer.py:266]: Move to if above

**Review by crewtool — CHANGES_REQUESTED** [review]

**Comment by crewtool** [comment]
Nicely done!! Small comment added

File: research/conditional/utils/conditional_trainer.py (hunk 1 above)
Comment (maciek-pioro) [reply research/conditional/utils/conditional_trainer.py:296]: TBH not really - we want the `with FSDP(...)` to run in all processes, but the logging to only be done by the logging process.

File: research/conditional/utils/conditional_trainer.py (hunk 2 above)
Comment (maciek-pioro) [reply research/conditional/utils/conditional_trainer.py:266]: Same as the previous comment :)

File: lizrd/core/distributed.py [hunk 3]
```diff
@@ -54,6 +54,7 @@ def wrap_in_fsdp(
         ),
         cpu_offload=CPUOffload(offload_params=offload_params),
         auto_wrap_policy=wrap_policy,
+        use_orig_params=True,
```
Comment (crewtool) [inline lizrd/core/distributed.py:57]: From:
https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel
`True is required to use torch.compile()`

File: research/conditional/utils/conditional_trainer.py (hunk 1 above)
Comment (crewtool) [reply research/conditional/utils/conditional_trainer.py:296]: Yup, found that in the docs. 
How about adding "rank0_only=True", and moving if self.is_logging_process: outside `for`

File: research/conditional/utils/conditional_trainer.py [hunk 4]
```diff
@@ -229,8 +255,58 @@ def calculate_loss_and_gradient(self, processed_batch: LLMBatch):
             "losses": losses,
         }
 
-    def _apply_gradient(self):
+    def maybe_report_gradient_norm(self, step: int):
+        if self.logging_interval_heavy > 0 and step % self.logging_interval_heavy == 0:
```
Comment (crewtool) [inline research/conditional/utils/conditional_trainer.py:259]:
```suggestion
        self.will_report_update_norm(step):
```

**Review by crewtool — APPROVED** [review]

File: research/conditional/utils/conditional_trainer.py (hunk 4 above)
Comment (crewtool) [reply research/conditional/utils/conditional_trainer.py:259]: or at least checking if model is a FSDP instance
